packages/dependency-extraction-webpack-plugin/lib/util.js

```diff
@@ -73,6 +77,14 @@ function defaultRequestToExternalModule( request ) {
 		// which forces @wordpress/interactivity imports to be hoisted to static imports.
 		return `module ${ request }`;
 	}
+
+	const isWordPressScript = Boolean( defaultRequestToExternal( request ) );
+
+	if ( isWordPressScript ) {
+		return new Error(
+			`Attempted to use WordPress script in a module: ${ request }`
+		);
+	}
```

Comment:
If we want to prevent bundling `@wordpress/*` modules by accident, then the `defaultRequestToExternalModule` should externalize everything `@wordpress/*` by default. It's then the user's responsibility to add them to the importmap, otherwise the runtime `import '@wordpress/x'` will fail.

That's an approach identical to scripts, only instead of ensuring that scripts are `wp_enqueue`d we need to ensure that modules are in importmap.

Reply:
If we really decide to treat `@wordpress/*` imports as an error by default, it's better to _throw_ the error rather than return it.

Reply:
> It's then the user's responsibility to add them to the importmap, otherwise the runtime import '@wordpress/x' will fail

I think we're not there yet, as there are no `@wordpress` packages available as modules except for `@wordpress/interactivity`.

Reply:
My understanding is that we can't use scripts from modules at all at this time. Adding them to the import map wouldn't help because they don't export anything, they're not modules.

---

We can throw the error here, that's fine. We'll need to `try/catch` and pass it to the callback in the plugin (https://github.com/WordPress/gutenberg/pull/57795/files#r1450176826)

Reply:
> My understanding is that we can't use scripts from modules at all at this time.

But even without this PR we will never use any actual WP script anyway. We'll either bundle a local `node_modules` package, or output an import of something that's expected to resolve to a ES module.

If we keep this, it feels intuitively wrong to me to piggy-pack on `defaultRequestToExternal`. If we want to forbid `@wordpress/*`, write it explicitly in `defaultRequestToExternalModule`. Also, the error message will be very confusing if it prevents an import of `lodash`, `react` or `moment`. These are common NPM libraries, not WordPress-specific.

Reply:
I share some of your reservations about this change. But I think that it's a good idea to provide this kind of feedback because during this scripts<->modules transition there's likely to be a lot of confusion and mistakes. This kind of change is intended to bring some clarity at an ideal time. It would help a lot of there were a post we could link to in the error message about the plans for this phase. @luisherranz is there anything public about scripts-modules compatibility we could link to?

---

This plugin _is_ specific to WordPress. It can be used outside of WordPress, but that's not its primary purpose. If we do use it outside of WordPress, we should set `useDefaults: false` and we wouldn't get any errors.

I piggy-backed on `defaultRequestToExternal` because the idea here is that if you're trying to use those scripts when compiling a Module for Wordpress, you're probably doing it wrong (at this time). `defaultRequestToExternal` is a pretty good approximation of what scripts are available and what we're likely to see used. There may be legacy scripts that will never be available as modules and maybe we should maintain a list of those going forward.

Some of those other modules probably work - `react`, `lodash`, `moment` - but I'd hate to have folks compiling a frontend module for multiple blocks and bundling those three dependencies in each one.

Reply:
Alright, please go forward then 🙂 I don't have any other feedback.